Mathlib/Topology/ContinuousFunction/FunctionalCalculus.lean

```diff
@@ -0,0 +1,676 @@
+    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
+    (spectrum R a).EqOn f g := by
+  rw [cfc_apply a f, cfc_apply a g] at h
+  have := (cfcSpec_closedEmbedding (show p a from ha) (R := R)).inj h
+  intro x hx
+  congrm($(this) ⟨x, hx⟩)
+
+lemma cfc_const (r : R) (ha : p a := by cfc_tac) :
+    cfc a (fun _ ↦ r) = algebraMap R A r := by
+  rw [cfc_apply a (fun _ : R ↦ r), ← AlgHomClass.commutes (cfcSpec ha (p := p)) r]
+  congr
+
+variable (R)
+
+lemma cfc_one (ha : p a := by cfc_tac) : cfc a (1 : R → R) = 1 :=
+  cfc_apply a (1 : R → R) ▸ map_one (cfcSpec (show p a from ha))
+
+lemma cfc_const_one (ha : p a := by cfc_tac) : cfc a (fun _ : R ↦ 1) = 1 :=
+  cfc_one R a
+
+@[simp]
+lemma cfc_zero : cfc a (0 : R → R) = 0 := by
+  by_cases ha : p a
+  · exact cfc_apply a (0 : R → R) ▸ map_zero (cfcSpec ha)
+  · rw [cfc_apply_of_not_predicate a ha]
+
+@[simp]
+lemma cfc_const_zero : cfc a (fun _ : R ↦ 0) = 0 :=
+  cfc_zero R a
+
+variable {R}
+
+lemma cfc_mul (f g : R → R)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
+    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
+    cfc a (fun x ↦ f x * g x) = cfc a f * cfc a g := by
+  by_cases ha : p a
+  · rw [cfc_apply a f, cfc_apply a g, ← map_mul, cfc_apply a _]
+    congr
+  · simp [cfc_apply_of_not_predicate a ha]
+
+lemma cfc_pow (f : R → R) (n : ℕ) (ha : p a := by cfc_tac)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)  :
+    cfc a (fun x ↦ (f x) ^ n) = cfc a f ^ n := by
+  rw [cfc_apply a f, ← map_pow, cfc_apply a _]
+  congr
+
+lemma cfc_add (f g : R → R)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
+    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
+    cfc a (fun x ↦ f x + g x) = cfc a f + cfc a g := by
+  by_cases ha : p a
+  · rw [cfc_apply a f, cfc_apply a g, ← map_add, cfc_apply a _]
+    congr
+  · simp [cfc_apply_of_not_predicate a ha]
+
+lemma cfc_smul {S : Type*} [SMul S R] [ContinuousConstSMul S R]
+    [SMulZeroClass S A] [IsScalarTower S R A] [IsScalarTower S R (R → R)]
+    (s : S) (f : R → R) (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) :
+    cfc a (fun x ↦ s • f x) = s • cfc a f := by
+  by_cases ha : p a
+  · rw [cfc_apply a f, cfc_apply a _]
+    simp_rw [← Pi.smul_def, ← smul_one_smul R s _]
+    rw [← map_smul]
+    congr
+  · simp [cfc_apply_of_not_predicate a ha]
+
+lemma cfc_const_mul (r : R) (f : R → R)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) :
+    cfc a (fun x ↦ r * f x) = r • cfc a f :=
+  cfc_smul a r f
+
+lemma cfc_star (f : R → R) : cfc a (fun x ↦ star (f x)) = star (cfc a f) := by
+  by_cases h : p a ∧ ContinuousOn f (spectrum R a)
+  · obtain ⟨ha, hf⟩ := h
+    rw [cfc_apply a f, ← map_star, cfc_apply a _]
+    congr
+  · obtain (ha | hf) := not_and_or.mp h
+    · simp [cfc_apply_of_not_predicate a ha]
+    · rw [cfc_apply_of_not_continuous a hf, cfc_apply_of_not_continuous, star_zero]
+      exact fun hf_star ↦ hf <| by simpa using hf_star.star
```

Comment:
All of these should either be stated for beta-reduced functions or for non-beta-reduced functions, but not a mix of both. Personally I think we need both versions, eg
```
lemma cfc_mul' (f g : R → R)
    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
    cfc a (f * g) = cfc a f * cfc a g := sorry

lemma cfc_mul (f g : R → R)
    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
    cfc a (fun x ↦ f x * g x) = cfc a (fun x ↦ f x) * cfc a (fun x ↦ g x) := sorry
```
What will happen if you don't do that is that your functions might accidentally get beta-reduced during a rewrite, forcing you to unreduce them again to be able to rewrite further.

Reply:
@ADedecker @dupuisf @lecopivo what do you think here? Personally, I prefer `f * g`, but went with the expanded versions because that's what `fun_prop` needed at the time. Currently, `fun_prop` still needs it this way, but that's actually only because of *existing* lemmas. That is, we have [Continuous.mul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Monoid.html#Continuous.mul), but no `Continuous.mul'` variant.

I guess the question is: can we uniformize across the library on a particular version, do we just need better automation, or is Yaël right that we need both versions everywhere (:cry: that's a lot of duplication).

(Also note, this isn't beta reduction, because that would need to be applied to an argument.)

Reply:
Sorry, I of course meant "<insert correct greek letter>-reduced", where here we may take <insert correct greek letter> = eta. 😀

Reply:
I always go for `f * g` just because it's cleaner, but I get that it may not be the easiest to rewrite with. I think the safest option is to expand everything for now because it matches more closely the existing API (even though that makes me a bit sad...). The outlier here is definitely composition though, which we usually do *not* eta-expand (e.g [Continuous.comp](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Basic.html#Continuous.comp) and [Continuous.comp'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Basic.html#Continuous.comp') are named in the reverse order).

Reply:
I've switched `cfc_comp` and `cfc_comp'`

Reply:
I personally prefer the reduced versions, I find that they're usually easier to work with.